hathor/websocket/streamer.py

```diff
@@ -0,0 +1,300 @@
+
+    def pauseProducing(self) -> None:
+        """Pause streaming. Called by twisted."""
+        self._paused = True
+
+    def stopProducing(self) -> None:
+        """Stop streaming. Called by twisted."""
+        self._stop = True
+        self.stop(False)
+
+    def resumeProducing(self) -> None:
+        """Resume streaming. Called by twisted."""
+        self._paused = False
+        self._run()
+
+    def _run(self) -> None:
+        """Run the streaming main loop."""
+        coro = self._async_run()
+        Deferred.fromCoroutine(coro)
+
+    async def _async_run(self):
+        """Internal method that runs the streaming main loop."""
+        t0 = self.reactor.seconds()
+
+        async for _type, data in self.search_iter:
+            if _type == 'address':
+                address_index, address = data
+                subscribed, errmsg = self.protocol.subscribe_address(address)
+
+                if not subscribed:
+                    self.send_message({
+                        'type': 'stream:history:error',
+                        'id': self.stream_id,
+                        'errmsg': f'Address subscription failed: {errmsg}'
+                    })
+                    self.stop(False)
+                    return
+
+                self.stats_sent_addresses += 1
+                self.send_message({
+                    'type': 'stream:history:address',
+                    'id': self.stream_id,
+                    'index': address_index,
+                    'address': address,
+                    'subscribed': subscribed,
+                })
+
+            elif _type == 'vertex':
+                tx = data
+                self.stats_sent_vertices += 1
+                self.send_message({
+                    'type': 'stream:history:vertex',
+                    'id': self.stream_id,
+                    'data': tx.to_json_extended(),
+                })
+
+            else:
+                assert False
+
+            self.stats_total_messages += 1
+            if self.stats_total_messages % self.log_every_n == 0:
+                self.protocol.log.info('websocket streaming statistics',
+                                       total_messages=self.stats_total_messages,
+                                       sent_vertices=self.stats_sent_vertices,
+                                       sent_addresses=self.stats_sent_addresses)
+
+            dt = self.reactor.seconds() - t0
+            if dt > self.max_seconds_locking_event_loop:
+                # Let the event loop run at least once.
+                await deferLater(self.reactor, 0, lambda: None)
+                t0 = self.reactor.seconds()
+
+            # The methods `pauseProducing()` and `stopProducing()` might be called during the
+            # call to `self.protocol.sendMessage()`. So both `_paused` and `_stop` might change
+            # during the loop.
+            if self._paused or self._stop:
+                break
```

Comment:
It doesn't make a difference. I put it at the end of the loop because the call to `sendMessage()` would be the one that can flip these flags but it can be anywhere.

Reply:
Couldn't twisted stop it by itself, and therefore set the flag, before?

Reply:
Clarifying: if Twisted closes the connection, it'll call `stopProducing()` which sets `self._stop = True`. In that case, the loop could try to send a message even after it should be stopped, no?

Reply:
> Does `stopProducing()` mean the producer should stop immediately or is it a suggestion for the producer to stop soon?

It means to stop immediately. I haven't tested but I think that any attempt to send more messages would fail.

Reply:
> Clarifying: if Twisted closes the connection, it'll call `stopProducing()` which sets `self._stop = True`. In that case, the loop could try to send a message even after it should be stopped, no?

You're right. I moved it to the beginning of the loop and also added a verification in the `else` block of the for. Nice catch! Thanks!

Reply:
Great, thanks!